### PR TITLE
feat(providers): TikTok provider

### PR DESCRIPTION
**Pull request by denkristoffer**
## Reasoning 💡

Adds TikTok provider. Their implementation is a bit funky, so there's a bit of customisation.

## Checklist 🧢

- [ ] Documentation
- [ ] Tests
- [x] Ready to be merged

### CODE REVIEW

File: src/providers/tiktok.ts
Comment (balazsorban44) [inline src/providers/tiktok.ts:48]: There is now a top-level option for setting timeout:

https://github.com/nextauthjs/next-auth/pull/3188

I would also suggest setting something more sensible, like 5-10 seconds.

File: src/providers/tiktok.ts
Comment (balazsorban44) [inline src/providers/tiktok.ts:25]:
```suggestion
      url: "https://open-api.tiktok.com/platform/oauth/connect",
```

File: src/providers/tiktok.ts
Comment (balazsorban44) [inline src/providers/tiktok.ts:28]:
```suggestion
      url: "https://open-api.tiktok.com/oauth/access_token",
```

File: src/providers/tiktok.ts
Comment (balazsorban44) [inline src/providers/tiktok.ts:45]:
```suggestion
      url: "https://open-api.tiktok.com/oauth/userinfo",
```

File: src/providers/tiktok.ts
Comment (balazsorban44) [inline src/providers/tiktok.ts:9]: Anything else that you can decipher from the returned `tikTokData` that could be useful for people here?

**Review by balazsorban44 — CHANGES_REQUESTED** [review]
Thank you! I added some suggestions.

Do note that we plan to release v4 to stable very soon, so it would be nice to get this merged before that.

Could you show some kind of proof of login that shows the returned `name`, `email`, `picture`, or any of those that are actually returned?

We could use that, later on, to refer to, when people are having issues.

Thank you!

**Comment by balazsorban44** [comment]
Gentle bump. We will release v4 next week, if this doesn't address my comments, we cannot merge it and therefore will be closed.

**Comment by balazsorban44** [comment]
An accompanying docs page should also be added at https://github.com/nextauthjs/docs/tree/main/docs/providers

**Comment by denkristoffer** [comment]
Thank you for the bump. I didn't mean to close this, I will try to get to this in time for v4. Rebasing on the `beta` branch.

**Comment by denkristoffer** [comment]
> Could you show some kind of proof of login that shows the returned name, email, picture, or any of those that are actually returned?
>
> We could use that, later on, to refer to, when people are having issues.

What are you looking for here? A demo website?

**Comment by balazsorban44** [comment]
Basically just a screenshot of the returned data by the `profile` callback. Either the JSON from a `console.log` or an image from the site. Example: https://github.com/nextauthjs/next-auth/pull/2848#issue-1008276523

File: src/providers/tiktok.ts
Comment (balazsorban44) [inline src/providers/tiktok.ts:56]: I think I'll update the `httpOptions` to expose `followRedirect` as well, so we hopefully can get rid of the `openid-client` `custom` import.

**Comment by balazsorban44** [comment]
If you merge `beta` into your branch now, https://github.com/nextauthjs/next-auth/pull/3287 exposes the `httpOptions`, so you can get rid of the above-mentioned import.

**Comment by balazsorban44** [comment]
I've tried to register an App, it had some caveats, that should be mentioned in the docs.

Would be appreciated if you opened a PR regarding this https://github.com/nextauthjs/next-auth/pull/2720#issuecomment-979500400

First of all, it seems like you cannot use `http`. even for local development. Second, my app registration has to go through a seemingly manual review process.

**Comment by denkristoffer** [comment]
Great! Docs PR is under way 🙂

**Comment by balazsorban44** [comment]
In that case, could you paste in the logged in result just so I (and future everyone) knows that this provider should work?

**Comment by denkristoffer** [comment]
I don't currently have a site using this provider, but I'll try to run it with the demo website later today.

**Comment by balazsorban44** [comment]
I don't expect an actual site. do try with the demo app. otherwise how did you even create the PR? 😀

**Comment by denkristoffer** [comment]
> I don't expect an actual site. do try with the demo app. otherwise how did you even create the PR? 😀

I created this for a personal project, but eventually decided against using TikTok for auth since their oauth implementation honestly seemed to change quickly.

I couldn't get the demosite working with the provider yesterday because of the oauth client complaining about an "Invalid URL". I will debug and try again tonight, hopefully.

**Comment by balazsorban44** [comment]
Unfortunately, if we cannot show proof of login, I will have to close this PR. I'll check back at the end of my day, and if works then, we can merge. Otherwise, I'll close this for now. We can revisit the idea at a later time after v4 has landed in stable.

Thank you for your contribution nonetheless.

**Comment by Oxicode** [comment]
anything news??

**Pull request closed without merging**
